File: src/test/java/nqueens/Master.java
package nqueens;

import abs.api.cwi.ABSFuture;
import abs.api.cwi.ActorSystem;
import abs.api.cwi.Guard;
import abs.api.cwi.LocalActor;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.function.Supplier;

public class Master extends LocalActor {
    int numWorkers;
    int priorities;
    int threshold;
    int size;

    LinkedList<Worker> workersSeq;

    private long t1 = System.currentTimeMillis();

    public Master(int numWorkers, int priorities, int threshold, int size) {
        this.numWorkers = numWorkers;
        this.priorities = priorities;
        this.threshold = threshold;
        this.size = size;
        workersSeq = new LinkedList<>();
        for (int i = 0; i < numWorkers; i++) {
            workersSeq.add(new Worker(this, threshold, size));
        }
    }
    public ABSFuture<List<int[]>> sendWork(int[] list, int depth, int priorities) {
        //Guard nonEmpty = Guard.convert(()->!workersSeq.isEmpty());
        //return spawn(nonEmpty, ()->{
            Worker w = workersSeq.pop();
            ABSFuture<List<int[]>> outcome = w.send(() -> w.nqueensKernelPar(list, depth, priorities));
            workersSeq.add(w);
            return outcome;
        //});
    }

    public ABSFuture<Void> finished(Worker w){
        workersSeq.add(w);
        return ABSFuture.done();
    }

    public ABSFuture<Void> init() {
        t1 = System.currentTimeMillis();
        System.out.println("COOP: Boardsize =" + size);
        int[] inArray = new int[size];
        ABSFuture<List<int[]>> resultF = this.send(() -> this.sendWork(inArray, 0, priorities));
        return getSpawn(resultF, result -> {
            System.out.println("Found " + result.size() + " solutions");
            System.out.println("-------------------------------- Program successfully completed! in " + (System.currentTimeMillis() - t1));
            this.send(() -> this.init());
            return ABSFuture.done();
        });
    }
}

File: src/test/scala/NQueens/nqueensconf/Master.scala
package NQueens.nqueensconf

import java.util.Objects

import com.ascoop.{ActorSystem, Future, TypedActor}


class Master(var numWorkers : Int,var priorities : Int,var solutionsLimit : Int,var threshold : Int,var size : Int) extends TypedActor with IMaster {

  private val workerSeq = for (_ <- 1 to numWorkers) yield {
    new Worker(this, threshold, size)
  }
  private final val workers = Iterator.continually(workerSeq).flatten

  var result: List[Array[Int]] = List()
  private var resultCounter: Int = 0
  private var t1 = System.currentTimeMillis()

  def success(solution: Array[Int]): Future[Void] = messageHandler {
      result = solution +: result
      resultCounter = resultCounter + 1
      if (Objects.equals(resultCounter, solutionsLimit)) {
        println(s"Found ${result.size} solutions")
        println("-------------------------------- Program successfully completed! in " + (System.currentTimeMillis() - t1))
        val inArray: Array[Int] = new Array[Int](0)
        result= List()
        resultCounter = 0
        t1 = System.currentTimeMillis()

        //delete these 2 lines to run only once
        this.sendWork(inArray, 0, priorities)
        Future.done()
        //ActorSystem.shutdown() uncomment this for termination after setting a fixed number of iterations or after one run.
      }
    else
        Future.done()
    }


  def sendWork(list: Array[Int], depth: Int, priorities: Int): Future[Void] = messageHandler {
//    println(s"Work $depth")
    val worker = workers.next()
    val f =worker.nqueensKernelPar(list, depth, priorities)
    Future.done()
  }

  def init = {
    println(s"NON-COOP: Boardsize = ${size.toString}, number of solutions should be ${solutionsLimit.toString}")
    val inArray: Array[Int] = new Array[Int](0)
    this.sendWork(inArray, 0, priorities)
  }
}
